packages/backend-common/src/reading/GcsUrlReader.test.ts

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ */
+
+import { ConfigReader, JsonObject } from '@backstage/config';
+import { getVoidLogger } from '../logging';
+import { ReadTreeResponseFactory } from './tree';
+import { GcsUrlReader } from './GcsUrlReader';
+import { UrlReaderPredicateTuple } from './types';
+
+describe('GcsUrlReader', () => {
+  const createReader = (config: JsonObject): UrlReaderPredicateTuple[] => {
+    return GcsUrlReader.factory({
+      config: new ConfigReader(config),
+      logger: getVoidLogger(),
+      treeResponseFactory: ReadTreeResponseFactory.create({
+        config: new ConfigReader({}),
+      }),
+    });
+  };
+
+  it('does not create a reader without the gcs field', () => {
+    const entries = createReader({
+      integrations: {},
+    });
+    expect(entries).toHaveLength(0);
+  });
+
+  it('creates a reader with credentials correctly configured', () => {
+    const entries = createReader({
+      integrations: {
+        gcs: [
+          {
+            privateKey: '--- BEGIN KEY ---- fakekey --- END KEY ---',
+            clientEmail: 'someone@example.com',
+          },
+          {
+            host: 'proxy.storage.cloud.google.com',
+            privateKey: '--- BEGIN KEY ---- fakekey2 --- END KEY ---',
+            clientEmail: 'someone2@example.com',
+          },
+        ],
+      },
+    });
+    expect(entries).toHaveLength(2);
+  });
+
+  it('does not create a reader if the privateKey is missing', () => {
+    const entries = createReader({
+      integrations: {
+        gcs: [
+          {
+            clientEmail: 'someone@example.com',
+          },
+        ],
+      },
+    });
+    expect(entries).toHaveLength(0);
+  });
+
+  it('does not create a reader if the clientEmail is missing', () => {
+    const entries = createReader({
+      integrations: {
+        gcs: [
+          {
+            privateKey:
+              '-----BEGIN PRIVATE KEY----- fakekey -----END PRIVATE KEY-----',
+          },
+        ],
+      },
+    });
+    expect(entries).toHaveLength(0);
+  });
+
+  it('predicates', () => {
+    const readers = createReader({
+      integrations: {
+        gcs: [
+          {
+            privateKey:
+              '-----BEGIN PRIVATE KEY----- fakekey -----END PRIVATE KEY-----',
+            clientEmail: 'someone@example.com',
+          },
+        ],
+      },
+    });
+    const predicate = readers[0].predicate;
+    expect(predicate(new URL('https://storage.cloud.google.com'))).toBe(true);
+    expect(
+      predicate(
+        new URL(
+          'https://storage.cloud.google.com/team1/service1/catalog-info.yaml',
+        ),
+      ),
+    ).toBe(true);
+    expect(predicate(new URL('https://storage2.cloud.google.com'))).toBe(false);
+    expect(predicate(new URL('https://cloud.google.com'))).toBe(false);
+    expect(predicate(new URL('https://google.com'))).toBe(false);
+    expect(predicate(new URL('https://a.example.com/test'))).toBe(false);
```

Comment:
This may sound a bit funny, but doesn't this always refer to one value?  Given that, the first `expect()` on row 100 will either _always_ pass _or_ fail but then literally _every other one_ will ALWAYS do the opposite, therefore sort of being less useful in context of providing testing?  It's not like it might just say test google.com to somehow make sure that's false, is sort of what I mean. These aren't being passed INTO a function to check it's return, they are validating the return which could never occur if the first test properly passes/fails.  Does that makes sense?  I may be missing something? :)

Reply:
Hi @adamdmharvey , what would be the best way to improve this test in your opinion? I'm open to ideas. Should I split each expect to it's own `it()`?